File: 221_maximal_square.java
class Solution {
    public int maximalSquare(char[][] matrix) {
        if (matrix.length == 0 || matrix[0].length == 0) return 0;

        int max = 0;
        int[][] f = new int[matrix.length][matrix[0].length];
        // initialize values with f[i][0] = matrix[i][0], f[0][j] = matrix[0][j]
        // since we use equal size as auxiliary space, our space complexity is O(m * n)
        for (int i = 0; i < matrix.length; i++) {
            f[i][0] = Character.getNumericValue(matrix[i][0]);
            max = Math.max(max, f[i][0]);
        }
        for (int j = 0; j < matrix[0].length; j++) {
            f[0][j] = Character.getNumericValue(matrix[0][j]);
            max = Math.max(max, f[0][j]);
        }

        // since we iterate each elements, our time complexity is O(m * n)
        for (int i = 1; i < matrix.length; i++) {
            for (int j = 1; j < matrix[0].length; j++) {
                if (matrix[i][j] == '1') {
                    // here we try to find the minimum length for f[i][j] can be
                    f[i][j] = Math.min(f[i - 1][j - 1], Math.min(f[i - 1][j], f[i][j - 1])) + 1;
                } else{
                    f[i][j] = 0;
                }
                max = Math.max(max, f[i][j]);
            }
        }
        return max * max;
    }
}

File: python/221_maximal_square.py
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        if not matrix or not matrix[0]: return 0

        largest = 0
        dp = [[0] * len(matrix[0]) for _ in range(len(matrix))]

        for i in range(len(matrix)):
            dp[i][0] = int(matrix[i][0])
            largest = max(largest, dp[i][0])

        for j in range(len(matrix[0])):
            dp[0][j] = int(matrix[0][j])
            largest = max(largest, dp[0][j])

        for i in range(1, len(matrix)):
            for j in range(1, len(matrix[0])):
                if matrix[i][j] == "1":
                    # since the minimum could be 0, and current matrix[i][j] == 1
                    # we need to plus one to have least 1 item
                    dp[i][j] = min(dp[i - 1][j - 1], dp[i - 1][j], dp[i][j - 1]) + 1
                else:
                    dp[i][j] = 0
                largest = max(largest, dp[i][j])

        return largest * largest


